src/libraries/System.Private.CoreLib/src/System/IO/BufferedStream.cs

```diff
@@ -262,20 +331,35 @@ public override async ValueTask DisposeAsync()
             finally
             {
                 _stream = null;
-                _buffer = null;
+
+                // Don't set the buffer to null, to avoid a NullReferenceException
+                // when users have a race condition in their code (i.e. they call
+                // Close when calling another method on Stream like Read).
+                if (!_actLikeFileStream)
+                {
+                    _buffer = null;
+                }
             }
         }
 
-        public override void Flush()
+        public override void Flush() => Flush(true);
+
+        internal void Flush(bool performActualFlush)
         {
             EnsureNotClosed();
 
             // Has write data in the buffer:
             if (_writePos > 0)
             {
-                FlushWrite();
-                Debug.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
-                return;
+                // EnsureNotClosed does not guarantee that the Stream has not been closed
+                // an example could be a call to fileStream.SafeFileHandle.Dispose()
+                // so to avoid getting exception here, we just ensure that we can Write before doing it
+                if (_stream!.CanWrite)
```

Comment:
Today what happens with FileStream if you dispose its SafeHandle and then flush?  Seems odd that we're adding overhead to the common case (another CanWrite check) to work around such misuse that will result in a meaningful exception.

Reply:
The problematic scenario is when you dispose the safe handle and then try to dispose to buffered stream, which tries to flush and fails.

```cs
[Fact]
public void DisposingBufferedStreamThatWrapsAFileStreamWhichHasBennClosedViaSafeFileHandleCloseDoesNotThrow()
{
    using (FileStream fs = new FileStream(GetTestFilePath(), FileMode.Create))
    {
        var bufferedStream = new BufferedStream(fs, 100);

        fs.SafeFileHandle.Dispose();

        bufferedStream.Dispose(); // must not throw
    }
}
```